### DISCUSSION TITLE
DataLoader but with strings not in GPU memory?

### DISCUSSION THREAD

**Discussion by turian** [Lightning-AI/pytorch-lightning#9340] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule]
I have a DataLoader reading from a DataSet, where each instance is `(x, y, string)`.

x and y are tensors I want loaded to GPU memory, which is great because pl does that for me.

However, I don't want the string in GPU memory, since it is only used for crossreferencing another table for scoring.

Or is there a workaround that I can use to do my string lookups?

**Comment by tchaton** [comment]
Hey @turian,

Only torch.tensor are moved to the GPU. A string won't be moved, so I am not sure to fully follow your questions.

Would you mind to share a script to better explain your problem ?

Best,
T.C

**Reply by turian** [reply]
To be more precise, I am returning:
```
x: Tensor, y: Tensor, z: Dict[str, Union[str, int]]
```
from my DataSet.

My fear was that `z` was going to the GPU and wasting time. Apparently that's not the case

**Reply by tchaton** [reply]
Oh I see. Yes, it would in this case as the default_collate used by the DataLoader will convert the int into tensors.

```
from torch.utils.data._utils.collate import default_collate

data = [{"indices": [1, 2]}, {"indices": [1, 2]}]
batch = default_collate(data)
print(batch) # {'indices': [tensor([1, 1]), tensor([2, 2])]}
```

You can override the `transfer_batch_to_device` from the LightningModule to prevent moving the `z` to the device.

Best,
T.C

**Reply by turian** [reply]
@tchaton basically, I want to maintain an index or a lookup to auxiliary information for each instance. What is the cleanest pattern for doing so?

Thank you

**Reply by tchaton** [reply]
The cleanest to me, would be to override the default_collate to make sure the indices aren't transformed into tensors and therefore aren't move automatically by Lightning on the device.

We already have similar mechanism in Lightning Flash: https://github.com/PyTorchLightning/lightning-flash

**Reply by turian** [reply]
@tchaton do you mind pointing me to which file that's in?

Do you mean:

https://github.com/PyTorchLightning/lightning-flash/blob/a1bece4361a7cb5449715bef9975696e96c5f9ae/flash/core/integrations/vissl/transforms/utilities.py

**Reply by tchaton** [reply]
Dear @turian,

Are you using Lightning Flash or another framework ?

If you are using your own code and DataLoader creation, then you can do `DataLoader(ds, collate_fn=my_custom_collate_fn)`, otherwise you can create an issue on Flash with a reproducible script there and we would help you out !

Best,
T.C

**Reply by turian** [reply]
@tchaton  I am just using lightning, not flash.
